File: cpp/allergies/allergies.cpp
#include "allergies.h"

namespace allergies {
    const std::map<std::string, int> allergy_test::map = {
            {"eggs",         0},
            {"peanuts",      1},
            {"shellfish",    2},
            {"strawberries", 3},
            {"tomatoes",     4},
            {"chocolate",    5},
            {"pollen",       6},
            {"cats",         7},
    };

    const std::vector<std::string> allergy_test::vec = {
            "eggs",
            "peanuts",
            "shellfish",
            "strawberries",
            "tomatoes",
            "chocolate",
            "pollen",
            "cats"
    };

    bool allergy_test::is_allergic_to(const std::string &food) const {
        auto postion = map.find(food);
        if (postion == map.end()) {
            return false;
        }

        auto result = mark & (1u << postion->second);

        return result != 0;
    }

    std::unordered_set<std::string> allergy_test::get_allergies() const {
        std::unordered_set<std::string> result;

        for (int i = 0; i < vec.size(); i++) {
            if ((mark & (1u << i)) != 0) {
                result.insert(vec[i]);
            }
        }

        return result;
    }
}

File: go/allergies/allergies.go
package allergies

import (
	"math/bits"
)

const testVersion = 1

var table = [...]string{
	"eggs",
	"peanuts",
	"shellfish",
	"strawberries",
	"tomatoes",
	"chocolate",
	"pollen",
	"cats",
}

// Allergies returns the food list for a certain number
func Allergies(number uint) []string {
	var result []string
	for i := 0; i < len(table); i++ {
		if number&(1<<uint(i)) != 0 {
			result = append(result, table[i])
		}
	}

	return result
}

// AllergicTo returns if a certain number is allergic to the food
func AllergicTo(number uint, food string) bool {
	for r := bits.TrailingZeros(number); r != bits.UintSize; {
		if food == table[r] {
			return true
		}

		number ^= 1 << uint(r)
		r = bits.TrailingZeros(number)
	}

	return false
}
